libraries/core_libs/network/src/tarcap/packets_handlers/latest/transaction_packet_handler.cpp

```diff
@@ -140,35 +151,125 @@ void TransactionPacketHandler::periodicSendTransactions(SharedTransactions &&tra
   }
 }
 
+std::vector<std::pair<std::shared_ptr<TaraxaPeer>, std::pair<SharedTransactions, std::vector<trx_hash_t>>>>
+TransactionPacketHandler::transactionsToSendToPeers(std::vector<SharedTransactions> &&transactions) {
+  // Main goal of the algorithm below is to send different transactions and hashes to different peers but still follow
+  // nonce ordering for single account and not send higher nonces without sending low nonces first
+  const auto accounts_size = transactions.size();
+  std::vector<std::pair<std::shared_ptr<TaraxaPeer>, std::pair<SharedTransactions, std::vector<trx_hash_t>>>>
+      peers_with_transactions_to_send;
+  auto peers = peers_state_->getAllPeers();
+
+  if (accounts_size) {
+    // trx_account_index keeps current account start index so that different peers will receive
+    // transactions from different accounts
+    uint32_t trx_account_index = 0;
+    for (const auto &peer : peers) {
+      if (!peer.second->syncing_) {
+        SharedTransactions peer_trxs;
+        std::vector<trx_hash_t> peer_trx_hashes;
+        bool trx_max_reached = false;
+        uint32_t trx_max_reached_index = 0;
+        auto trx_account_index_iterator = trx_account_index;
+
+        while (true) {
+          // Iterate over transactions from single account
+          for (auto const &trx : transactions[trx_account_index_iterator]) {
+            auto trx_hash = trx->getHash();
+            if (peer.second->isTransactionKnown(trx_hash)) {
+              continue;
+            }
+            // If max number of transactions to be sent is already reached include hashes to be sent
+            if (trx_max_reached) {
+              peer_trx_hashes.push_back(trx_hash);
+              if (peer_trx_hashes.size() == kMaxHashesInPacket) {
+                // If both transactions and hashes reached max nothing to do for this peer, break
+                break;
+              }
+            } else {
+              peer_trxs.push_back(trx);
+              if (peer_trxs.size() == kMaxTransactionsInPacket) {
+                // Max number of transactions reached, save trx_max_reached_index for next peer to continue to avoid
+                // sending same transactions to multiple peers
+                trx_max_reached = true;
+                trx_max_reached_index = (trx_account_index_iterator + 1) % accounts_size;
+              }
+            }
+          }
```

Comment:
Looks like it is better to move this into a separate function that will be collecting transactions and hashes for the peer starting from some index and return the index for the next peer to start from